EpsilonAuxBMS/Src/SetContactor/StateFunctions.c

```diff
@@ -0,0 +1,357 @@
+}
+
+void blocked(SetContactorStateInfo* stateInfo)
+{
+    const uint8_t commonSense = !HAL_GPIO_ReadPin(COMMON_SENSE_GPIO_Port, COMMON_SENSE_Pin);
+    const uint8_t chargeSense = !HAL_GPIO_ReadPin(CHARGE_SENSE_GPIO_Port, CHARGE_SENSE_Pin);
+    const uint8_t dischargeSense = !HAL_GPIO_ReadPin(DISCHARGE_SENSE_GPIO_Port, DISCHARGE_SENSE_Pin);
+
+    // If orion isn't overiding the contactors, but they are supposed to have been set, but aren't
+    // , this is an error condition and all contactors should be disconnected.
+    if (stateInfo->prevState > COMMON_ENABLE_CHECK &&
+            (!commonSense || chargeDisconnectedDuringContactorSetting(chargeSense, stateInfo->hasChargeBeenSet) ||
+             dischargeDisconnectedDuringContactorSetting(dischargeSense, stateInfo->hasDischargeBeenSet)))
+    {
+        disconnectContactors(1);
+        stateInfo->state = CONTACTOR_DISCONNECTED;
+    }
+    else if (orionStatus.batteryVoltagesInRange)
+    {
+        if (orionStatus.chargeContactorOverriden)
+        {
+            stateInfo->hasChargeBeenSet = 0;
+        }
+
+        if (orionStatus.dischargeContactorOverriden)
+        {
+            stateInfo->hasDischargeBeenSet = 0;
+        }
+
+        if (osMutexWait(orionStatus.orionStatusMutex, 0) != osOK)
+        {
+            return;
+        }
+
+        orionStatus.chargeContactorOverriden = 0;
+        orionStatus.dischargeContactorOverriden = 0;
+
+        osMutexRelease(orionStatus.orionStatusMutex);
+        stateInfo->state = FIRST_CHECK;
+    }
+}
+
+const uint8_t dischargeDisconnectedDuringContactorSetting(const uint8_t dischargeSense, uint8_t hasDischargeBeenSet)
+{
+    return (!orionStatus.dischargeContactorOverriden && hasDischargeBeenSet && !dischargeSense);
+}
+
+const uint8_t chargeDisconnectedDuringContactorSetting(const uint8_t chargeSense, uint8_t hasChargeBeenSet)
+{
+    return (!orionStatus.chargeContactorOverriden && hasChargeBeenSet && !chargeSense);
+}
+
+const uint8_t canTurnOnDishargeContactor(const uint8_t dischargeSense)
+{
+    return (orionStatus.allowDischarge && orionStatus.dischargeContactorOverriden && !dischargeSense);
+}
+
+const uint8_t canTurnOnChargeContactor(const uint8_t chargeSense)
+{
+    return (orionStatus.allowCharge && orionStatus.chargeContactorOverriden && !chargeSense);
+}
+
+const uint8_t dischargeDisconnectedUnexpectedly(const uint8_t dischargeSense, const uint8_t isDischargeTurningOn)
+{
+    return (orionStatus.allowDischarge && !dischargeSense && !isDischargeTurningOn && !orionStatus.dischargeContactorOverriden);
+}
+
+const uint8_t chargeDisconnectedUnexpectedly(const uint8_t chargeSense, const uint8_t isChargeTurningOn)
+{
+    return (orionStatus.allowCharge && !chargeSense && !isChargeTurningOn && !orionStatus.chargeContactorOverriden);
+}
+
+const uint8_t safeToSetContactor(ContactorsSettingState state)
+{
+    return (!orionStatus.batteryVoltagesInRange && !auxStatus.startUpSequenceDone && state != CONTACTOR_DISCONNECTED);
+}
```

Comment:
It might be a good idea to move these functions to their own file. @bill-luu What do you think?

Reply:
Eh, it's not a big deal unless Bill cares

Reply:
If you guys aren't too tired of this, if we put each state in it's own file, then functions that are exclusively used in that a state can go into that file. (e.g `chargeDisconnectedUnexpectedly` and `dischargeDisconnectedUnexpectedly` can go in `idle.c`), and common ones can stay here. 

I notice that `safeToSetContactor` is the only function _not_ being used by any of the states but instead just in `SetContactorsTask.c`. If thats the case it's ok to just move it over there!